### PR TITLE
WIP: Drag to select

### PR DESCRIPTION
**Pull request by Eddie-CooRo**
Add drag to select support for dayRangeCalendar

**Todo**:
- [x] Implement drag to select functionality
- [x] Fix onDisabledDayError functionality
- [x] Fix bug when after selecting the range completely at one side of a disabled day, we can't reset our selection either select the first day at the opposite side
- [x] Improve behavior when the user selects one day by a simple click and then starts dragging from the selected day (intended behavior is selecting the second state instead of changing the first one)
- [x] Fix bug reported by [this](https://github.com/Kiarash-Z/react-persian-calendar-date-picker/pull/2#issuecomment-488601331) comment (second gif)
- [ ] Add support for touchscreens
- [ ] Fix functionality for single day date picker use case
- [ ] Find a better workaround for closing Calendar after selection is completed (I'm thinking of something like using a callback named onSelectCompleted or so)
- [ ] Document these changes

related to issues: #1

### CODE REVIEW

**Comment by Eddie-CooRo** [comment]
I can use some help in writing documentation :)

**Comment by Eddie-CooRo** [comment]
Also, let me know what you think about accepting a flag named enableDragToSelect or something like that in order to enable this feature.
Personally, I think since there is already a lot of abstraction implemented in this library, it's better to have this feature out of the box and without a flag, but it would be nice to have a way for bailing out (i.e having the enableDragToSelect flag, but setting it to true by default)

**Comment by Kiarash-Z** [comment]
@Eddie-CooRo Thanks for your contribution. there are a few issues:
1- It's working on mobile

2- If you select a starting day by a simple click and then drag from the day you just selected, it will get a little bit glitchy:
May-02-2019 13-26-48

2- Follow these steps to reproduce:
- Start a dragging 
- Drag to the end of  days and on a "-blank" day(don't release your mouse. On the following .gif file I made the blank days' background red for clarification)
- Release your mouse click on a blank day
- Now hover the calendar days

dfs

These sort of contributions warm my heart. Keep up the good work 🎊

**Comment by Kiarash-Z** [comment]
> Also, let me know what you think about accepting a flag named enableDragToSelect or something like that in order to enable this feature.
> Personally, I think since there is already a lot of abstraction implemented in this library, it's better to have this feature out of the box and without a flag, but it would be nice to have a way for bailing out (i.e having the enableDragToSelect flag, but setting it to true by default)

I agree. I think it would be better to have this flag and enable it by default(default value set to true)

**Comment by Kiarash-Z** [comment]
There's one more issue. Just check the times Calendar gets rerendered. It seems there are unnecessary rerenders.

File: src/Calendar.js [hunk 1]
```diff
@@ -40,6 +39,19 @@ const Calendar = ({
     cycleCount: 1,
     activeDate: null,
   });
+  const [firstSelectedDate, setFirstSelectedDate] = useState(null);
+  const [secondSelectedDate, setSecondSelectedDate] = useState(null);
+  const [currentlyDragging, setCurrentlyDragging] = useState(false);
```
Comment (Kiarash-Z) [inline src/Calendar.js]: I think this should be rather a computed state than a normal state value.

Comment (Eddie-CooRo) [reply src/Calendar.js]: Do you mean computing them based on initial selectedDayRange values? In this case, I agree with you and I'll fix them

**Comment by Eddie-CooRo** [comment]
> There's one more issue. Just check the times Calendar gets rerendered. It seems there are unnecessary rerenders.

Yes, I haven't considered performance at all :)
I'm currently somewhere between "make it work/make it right" steps, after finishing with the functionality we can work on "making it fast"

File: src/Calendar.js (hunk 1 above)
Comment (Kiarash-Z) [reply src/Calendar.js]: That one as well, but I mean `currentlyDragging` shouldn't be a state value. It gotta be a computed variable based on `firstSelectedDate` and `secondSelectedDate` and **the state of the mouse down/up**. It looks like **it won't trigger a necessary render by the change of its value**. We will discuss this on the "making it fast" phase😊

Comment (Eddie-CooRo) [reply src/Calendar.js]: Yeah!!! you're right. I haven't noticed it (yet😂)
we can easily use "useRef" in order to track which end of selectedRange is getting dragged.

**Comment by Eddie-CooRo** [comment]
> 1- It's not working on mobile(actually not working on touchscreens because of mouse-based events. there are [pointer events](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events) for touchscreens)

I can't get how pointer-events could help us with supporting touchscreens.
Currently, I've added some supported touch handlers to day items, but there is no onMouseEnter equivalent for touch screens. I've tried to implement it with onTouchMove but it needed too much effort and I'm so sleepy now :)

I think we can merge this feature event without onTouchEnter event. currently, it works just as before on touchscreens and supports drag-to-select on pointer-based screens

**Comment by Kiarash-Z** [comment]
> > 1- It's not working on mobile(actually not working on touchscreens because of mouse-based events. there are [pointer events](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events) for touchscreens)
> 
> I can't get how pointer-events could help us with supporting touchscreens.
> Currently, I've added some supported touch handlers to day items, but there is no onMouseEnter equivalent for touch screens. I've tried to implement it with onTouchMove but it needed too much effort and I'm so sleepy now :)
> 
> I think we can merge this feature event without onTouchEnter event. currently, it works just as before on touchscreens and supports drag-to-select on pointer-based screens

Thanks you Eddie for your efforts. Yeah, the trick is to use `onTouchMove`. Maybe this dnd is too much for now, but If it's gonna get merged we need to have the touchscreens support. Because it's whether draggable or not. We can't just add this whole of code and then document "Not working on mobile".

**Comment by Kiarash-Z** [comment]
By the way, I checked both mentioned bugs again, and they weren't fixed 🤔Can you check them again, please?

**Comment by Eddie-CooRo** [comment]
Yeah, I checked both of them and they seem fixed to me :)
Maybe I got something wrong :thinking: 
Here are corresponding gifs:

2- 

3-

**Comment by Eddie-CooRo** [comment]
> If it's gonna get merged we need to have the touchscreens support. 
Ok. I'll work on it :)

**Comment by Kiarash-Z** [comment]
> Yeah, I checked both of them and they seem fixed to me :)
> Maybe I got something wrong 🤔

Sorry, my bad. I hadn't pulled the latest changes, but the second one seems a little bit broken. DatePicker crashes after dragging the mouse over the blank days. Maybe your local changes haven't got committed yet. We will discuss this on your next commit.

**Comment by Eddie-CooRo** [comment]
> > Yeah, I checked both of them and they seem fixed to me :)
> > Maybe I got something wrong thinking
> 
> Sorry, my bad. I hadn't pulled the latest changes, but the second one seems a little bit broken. DatePicker crashes after dragging the mouse over the blank days. Maybe your local changes haven't got committed yet. We will discuss this on your next commit.

No problem.
Actually, I've committed everything so far, can you describe the steps to reproduce for getting DatePicker crashed? It looks ok in my laptop, maybe there are some compatibility issues between browsers?

File: src/Calendar.js [hunk 2]
```diff
@@ -221,6 +221,20 @@ const Calendar = ({
     return allDays;
   };
 
+  function onDaySelect(date) {
+    dispatchSelectedDays({ type: 'SELECT', date });
```
Comment (Kiarash-Z) [inline src/Calendar.js]: You're passing `isStandard` argument to `onDaySelect`, but it hasn't used in the function. By the way, I think maybe using ES6 syntax would be better for these functions.

Comment (Eddie-CooRo) [reply src/Calendar.js]: Yeah, it's so much better. Except that eslint complains about returning nothing in an arrow function :smile: 
Also, about the isStandard, I passed it to every handler function to keep the signature consistent between them.

**Comment by Kiarash-Z** [comment]
> No problem.
> Actually, I've committed everything so far, can you provide steps to reproduce for getting DatePicker crashed? It looks ok in my laptop, maybe there are some compatibility issues between browsers?

The s2r are the same. here's a gif:
error

**Comment by Eddie-CooRo** [comment]
> The s2r are the same. here's a gif:

This doesn't happen to me :|
Can you show me the error it gives to you?

**Comment by Eddie-CooRo** [comment]
Oops. I found the error, give me some minutes :)

File: src/Calendar.js (hunk 2 above)
Comment (Kiarash-Z) [reply src/Calendar.js]: Why would it complain? Just replaced them with ES6 arrow functions with no problem. Maybe you have a global config for ESLint. Make sure you run `yarn lint` to check style guides. I think `isStandard` can be passed along with date to `onDaySelect` and alike functions like this:
```
onDaySelect({ ...dayItem, isStandard });
onDayEnter({ ...dayItem, isStandard });
```
and then we don't have to define `isStandard` parameter in `onDayEnter`.

**Comment by Kiarash-Z** [comment]
Thanks for 

> Oops. I found the error, give me a couple of minutes :)

Thanks for the fix. Found a new issue. the dragging seems broken on touchscreens:
err

**Comment by Kiarash-Z** [comment]
One more thing.`firstSelectedDay` and `secondSelectedDay` should be derived from `selectedDayRange`. They're not our state value. We gotta send the new value for them through `onChange` prop. It seems you've removed the commented code for them

**Comment by Kiarash-Z** [comment]
Hi Eddie,
If you're busy these days, I can close this one for now, and you can submit a new PR when you're a little bit free.

**Pull request closed without merging**